### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
total_players = len(purchase_data['SN'].unique())
player_total = pd.DataFrame({"Total Players" : [total_players]})
player_total

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
unique_items = len(purchase_data['Item ID'].unique())
average_price = round(purchase_data['Price'].mean(),2)
num_purchases = len(purchase_data['Item ID'])
total_rev = round(purchase_data['Price'].sum(),2)
total_revnew = ("${:,.2f}".format)(total_rev)
average_pricenew = ("${:,.2f}".format)(average_price)

In [4]:
purchasing_analysis = pd.DataFrame({'Number of Unique Items' : [unique_items], 'Average Price': [average_pricenew], 'Number of Purchases' : [num_purchases], 'Total Revenue' : [total_revnew]})
purchasing_analysis

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
gender_df = purchase_data[['SN', 'Gender']]
gender_dfgroup = gender_df.groupby('Gender')['SN'].nunique()
gendergroup_df = pd.DataFrame({'Total Count': gender_dfgroup})

In [6]:
gender_total = pd.to_numeric(gendergroup_df['Total Count'])

In [7]:
gen_percent = round(gendergroup_df['Total Count']/(total_players)*100,2).astype(str) + '%'

In [8]:
gender_demographic = pd.merge(gendergroup_df, gen_percent, on = 'Gender')
gender_demographic = gender_demographic.rename(columns={'Total Count_x': 'Total Count', 'Total Count_y' : 'Percentage of Players'})

In [9]:
gender_demographic.sort_values('Percentage of Players', ascending = False)

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [10]:
gender_analysisdf = purchase_data[['Gender', 'Item ID']]

In [11]:
gender_analysisgroup = gender_analysisdf.groupby(['Gender'])
gender_purchasetotal = gender_analysisgroup.count()
gender_purchasetotal = gender_purchasetotal.rename(columns= {'Item ID' : 'Purchase Count'})

In [12]:
gender_purchasedf = purchase_data[['SN', 'Gender', 'Price']]

In [13]:
genderpurchase_group = gender_purchasedf.groupby(['Gender'])
gender_purchaseave = round(genderpurchase_group.mean(),2)
gender_purchaseave = gender_purchaseave.rename(columns={'Price' : 'Average Purchase Price'})

In [14]:
gender_purchasevalue = genderpurchase_group.sum()
gender_purchasevalue = gender_purchasevalue.rename(columns= {'Price' : 'Total Purchase Value'})

In [15]:
gen_purchase_analysis = pd.merge(gender_purchasetotal, gender_purchaseave, on= 'Gender')

In [16]:
gen_purchase_analysisdf = pd.merge(gen_purchase_analysis, gender_purchasevalue, on= 'Gender')
ave_purchase_person = round(gender_purchasevalue['Total Purchase Value']/gender_demographic ['Total Count'],2)
gen_purchase_analysisdf['Average Purchase Total per Person by Gender'] = ave_purchase_person
gen_purchase_analysisdf ['Average Purchase Price'] = gen_purchase_analysisdf ['Average Purchase Price'].astype(float).map('${:,.2f}'.format)
gen_purchase_analysisdf ['Total Purchase Value'] = gen_purchase_analysisdf ['Total Purchase Value'].astype(float).map('${:,.2f}'.format)
gen_purchase_analysisdf ['Average Purchase Total per Person by Gender'] = gen_purchase_analysisdf ['Average Purchase Total per Person by Gender'].astype(float).map('${:,.2f}'.format)
gen_purchase_analysisdf

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value  \
Gender                                       
Female                             $361.94   
Male                             $1,967.64   
Other / Non-Disclosed               $50.19   

                      Average Purchase Total per Person by Gender  
Gender                                                             
Female                                                      $4.47  
Male                                                        $4.07  
Other / Non-Disclosed                                       $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [17]:
print(purchase_data['Age'].max())
bins = [0, 9, 14, 19, 24, 29, 34, 39, 50]
age_labels = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

45


In [18]:
purchase_data['Age Bracket'] = pd.cut(purchase_data['Age'], bins, labels = age_labels, include_lowest=True)

In [19]:
age_df = purchase_data[['SN', 'Item ID', 'Age Bracket']]
age_dfgroup = age_df.groupby('Age Bracket')['SN'].nunique()
agegroup_df = pd.DataFrame({'Total Count': age_dfgroup})
age_percent = round(agegroup_df['Total Count']/(total_players)*100,2).astype(str) + '%'
agegroup_df['Percentage of Players'] = age_percent
agegroup_df

Total Count Percentage of Players
Age Bracket                                   
<10                   17                 2.95%
10-14                 22                 3.82%
15-19                107                18.58%
20-24                258                44.79%
25-29                 77                13.37%
30-34                 52                 9.03%
35-39                 31                 5.38%
40+                   12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [20]:
age_purchase = purchase_data[['Age Bracket', 'Item ID']]
age_purchasegroup = age_purchase.groupby(['Age Bracket'])
agepurchase_total = age_purchasegroup.count()
agepurchase_total = agepurchase_total.rename(columns={'Item ID': 'Purchase Count'})

In [21]:
agepurchase_df = purchase_data[['SN','Age Bracket', 'Price']]
agepurchase_dfgroup = agepurchase_df.groupby(['Age Bracket'])

In [22]:
agepurchase_ave = round(agepurchase_dfgroup.mean(),2)
agepurchase_ave = agepurchase_ave.rename(columns = {'Price' : 'Average Purchase Price'})

In [23]:
agepurchase_value = round(agepurchase_dfgroup.sum(),2)
agepurchase_value = agepurchase_value.rename(columns={'Price': 'Total Purchase Value'})


In [24]:
agepurchase_analysis = pd.merge(agepurchase_total, agepurchase_ave, on = 'Age Bracket')

In [25]:
agepurchase_summary = pd.merge(agepurchase_analysis, agepurchase_value, on='Age Bracket')
ave_purchase_age = round(agepurchase_value['Total Purchase Value']/agegroup_df['Total Count'],2)
agepurchase_summary['Average Purchase Total per Person by Age Group'] = ave_purchase_age
agepurchase_summary['Average Purchase Price'] = agepurchase_summary['Average Purchase Price'].astype(float).map('${:,.2f}'.format)
agepurchase_summary['Total Purchase Value'] = agepurchase_summary['Total Purchase Value'].astype(float).map('${:,.2f}'.format)
agepurchase_summary['Average Purchase Total per Person by Age Group'] = agepurchase_summary['Average Purchase Total per Person by Age Group'].astype(float).map('${:,.2f}'.format)
agepurchase_summary

Purchase Count Average Purchase Price Total Purchase Value  \
Age Bracket                                                               
<10                      23                  $3.35               $77.13   
10-14                    28                  $2.96               $82.78   
15-19                   136                  $3.04              $412.89   
20-24                   365                  $3.05            $1,114.06   
25-29                   101                  $2.90              $293.00   
30-34                    73                  $2.93              $214.00   
35-39                    41                  $3.60              $147.67   
40+                      13                  $2.94               $38.24   

            Average Purchase Total per Person by Age Group  
Age Bracket                                                 
<10                                                  $4.54  
10-14                                                $3.76  
15-19                                                $3.86  
20-24                                                $4.32  
25-29                                                $3.81  
30-34                                                $4.12  
35-39                                                $4.76  
40+                                                  $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [26]:
topspender_df = purchase_data[['SN','Item ID']]
topspender_group = topspender_df.groupby(['SN'])

In [27]:
snpurchasetotal = topspender_group.count()
snpurchasetotal = snpurchasetotal.rename(columns={'Item ID': 'Purchase Count'})

In [28]:
topspender_df2 = purchase_data[['SN','Price']]
topspenderdf_group = topspender_df2.groupby(['SN'])

In [30]:
topspender_ave = round(topspenderdf_group.mean(),2)
topspender_ave = topspender_ave.rename(columns= {'Price': 'Average Purchase Price'})

In [31]:
topspender_value = round(topspenderdf_group.sum(),2)
topspender_value = topspender_value.rename(columns={'Price': 'Total Purchase Value'})

In [32]:
topspender_analysis = pd.merge(snpurchasetotal, topspender_ave, on = 'SN')

In [83]:
topspender_summary = pd.merge(topspender_analysis, topspender_value, on = 'SN')

In [84]:
topspender_summary['Average Purchase Price'] = topspender_summary['Average Purchase Price'].astype(float).map('${:,.2f}'.format)
topspender_summary['Total Purchase Value'] = topspender_summary['Total Purchase Value'].astype(float).map('${:,.2f}'.format)
totalpurchase = topspender_summary['Total Purchase Value'].str.strip('$').astype(float)
topspender_summary2 = topspender_summary.iloc[totalpurchase.argsort()[::-1]]
topspender_summary2.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [63]:
popular_items = purchase_data[['Item ID', 'Item Name']]
popular_itemsgroup = popular_items.groupby(['Item ID'])

In [64]:
popularitem_total = popular_itemsgroup.count()
popularitem_total = popularitem_total.rename(columns={'Item Name': 'Purchase Count'})

In [65]:
popular_itemsdf = purchase_data[['Item ID', 'Price']]
popularitemsdf_group = popular_itemsdf.groupby(['Item ID'])

In [66]:
popularitem_ave = round(popularitemsdf_group.mean(),2)
popularitem_ave = popularitem_ave.rename(columns = {'Price': 'Item Price'})

In [67]:
popularitem_value = round(popularitemsdf_group.sum(),4)
popularitem_value = popularitem_value.rename(columns={'Price': 'Total Purchase Value'})

In [68]:
popularitem_table = pd.merge(popularitem_total, popularitem_ave, on= 'Item ID')

In [69]:
popularitem_summary2 = pd.merge(popularitem_table, popularitem_value, on= 'Item ID')

In [70]:
popularitem_summarynew = purchase_data[['Item ID', 'Item Name']]
popularitem_summarynew = popularitem_summarynew.drop_duplicates()
popularitem_analysis = pd.merge(popularitem_summarynew, popularitem_summary2, on = 'Item ID')

In [71]:
popularitem_analysisfin= popularitem_analysis.set_index(['Item ID', 'Item Name'])
popularitem_analysisfin['Item Price'] = popularitem_analysisfin['Item Price'].astype(float).map('${:,.2f}'.format)
popularitem_analysisfin['Total Purchase Value'] = popularitem_analysisfin['Total Purchase Value'].astype(float).map('${:,.2f}'.format)
popularitem_analysisfin.sort_values('Purchase Count', ascending=False).head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [80]:
profitable_items = popularitem_analysis

In [81]:
profitable_items= profitable_items.set_index(['Item ID', 'Item Name'])

In [82]:
profitable_items['Item Price'] = profitable_items['Item Price'].astype(float).map('${:,.2f}'.format)
profitable_items['Total Purchase Value'] = profitable_items['Total Purchase Value'].astype(float).map('${:,.2f}'.format)
newprofitable = profitable_items['Total Purchase Value'].str.strip('$').astype(float)
profitable_items2 = profitable_items.iloc[newprofitable.argsort()[::-1]]
profitable_items2.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

SUMMARY:

Majority of the players for Heroes of Pymoli are males which consists of a total of 
84.03%. They are also responsible for 83.59% of the total purchases.

The fantasy game is popular among the age bracket of 20-24yr olds with total purchase count of 46.79%.

Items such as Final Critic and Oathbreaker, Last Hope of the Breaking Storm, Nirvana, and Fiery Glass Crusader are in both most popular and profitable items.